cpp/include/cuvs/neighbors/ann_mg.hpp

```diff
@@ -0,0 +1,1417 @@
+                            const std::string& filename);
+
+  void build(const cuvs::neighbors::mg::nccl_clique& clique,
+             const cuvs::neighbors::index_params* index_params,
+             raft::host_matrix_view<const T, int64_t, row_major> index_dataset);
+
+  void extend(const cuvs::neighbors::mg::nccl_clique& clique,
+              raft::host_matrix_view<const T, int64_t, row_major> new_vectors,
+              std::optional<raft::host_vector_view<const IdxT, int64_t>> new_indices);
+
+  void search(const cuvs::neighbors::mg::nccl_clique& clique,
+              const cuvs::neighbors::search_params* search_params,
+              raft::host_matrix_view<const T, int64_t, row_major> queries,
+              raft::host_matrix_view<IdxT, int64_t, row_major> neighbors,
+              raft::host_matrix_view<float, int64_t, row_major> distances,
+              cuvs::neighbors::mg::sharded_merge_mode merge_mode,
+              int64_t n_rows_per_batch) const;
+
+  void serialize(raft::resources const& handle,
+                 const cuvs::neighbors::mg::nccl_clique& clique,
+                 const std::string& filename) const;
+
+ private:
+  void sharded_search_with_direct_merge(const cuvs::neighbors::mg::nccl_clique& clique,
+                                        const cuvs::neighbors::search_params* search_params,
+                                        raft::host_matrix_view<const T, int64_t, row_major> queries,
+                                        raft::host_matrix_view<IdxT, int64_t, row_major> neighbors,
+                                        raft::host_matrix_view<float, int64_t, row_major> distances,
+                                        int64_t n_rows_per_batch,
+                                        int64_t n_rows,
+                                        int64_t n_cols,
+                                        int64_t n_neighbors,
+                                        int64_t n_batches) const;
+
+  void sharded_search_with_tree_merge(const cuvs::neighbors::mg::nccl_clique& clique,
+                                      const cuvs::neighbors::search_params* search_params,
+                                      raft::host_matrix_view<const T, int64_t, row_major> queries,
+                                      raft::host_matrix_view<IdxT, int64_t, row_major> neighbors,
+                                      raft::host_matrix_view<float, int64_t, row_major> distances,
+                                      int64_t n_rows_per_batch,
+                                      int64_t n_rows,
+                                      int64_t n_cols,
+                                      int64_t n_neighbors,
+                                      int64_t n_batches) const;
+
+  distribution_mode mode_;
+  int num_ranks_;
+  std::vector<ann_interface<AnnIndexType, T, IdxT>> ann_interfaces_;
+};
+
+/// \defgroup ann_mg_cpp_index_build ANN MG index build
+
+/// \ingroup ann_mg_cpp_index_build
+/**
+ *
```

Comment:
Please write descriptions for all public API functions. The usage examples are great, but docs look much cleaner when there's nice descriptions of the expected behavior to accompany them.